src/manager_service.py

```diff
@@ -0,0 +1,146 @@
+_SERVICE_SETUP_ERROR_MESSAGE = "Unable to enable or start the github-runner-manager application"
+
+logger = logging.getLogger(__name__)
+
+
+def setup(state: CharmState, app_name: str, unit_name: str) -> None:
+    """Set up the github-runner-manager service.
+
+    Args:
+        state: The state of the charm.
+        app_name: The Juju application name.
+        unit_name: The Juju unit.
+    """
+    config_file = _setup_config_file(state, app_name, unit_name)
+    _setup_service_file(config_file)
+    _enable_service()
+
+
+def install_package() -> None:
+    """Install the GitHub runner manager package.
+
+    Raises:
+        RunnerManagerApplicationInstallError: Unable to install the application.
+    """
+    logger.info("Upgrading pip")
+    try:
+        execute_command(["python3", "-m", "pip", "install", "--upgrade", "pip"])
+    except SubprocessError as err:
+        raise RunnerManagerApplicationInstallError(_INSTALL_ERROR_MESSAGE) from err
+
+    logger.info("Uninstalling previous version of packages")
+    try:
+        execute_command(["python3", "-m", "pip", "uninstall", GITHUB_RUNNER_MANAGER_PACKAGE])
+        execute_command(["python3", "-m", "pip", "uninstall", JOB_MANAGER_PACKAGE])
+    except SubprocessError:
+        logger.info(
+            "Unable to uninstall existing packages, likely due to previous version not installed"
+        )
+
+    try:
+        # Use `--prefix` to install the package in a location (/usr) all user can use and
+        # `--ignore-installed` to force all dependencies be to installed under /usr.
+        execute_command(
+            [
+                "python3",
+                "-m",
+                "pip",
+                "install",
+                "--prefix",
+                "/usr",
+                "--ignore-installed",
+                GITHUB_RUNNER_MANAGER_PACKAGE_PATH,
+                JOB_MANAGER_PACKAGE_PATH,
+            ]
+        )
+    except SubprocessError as err:
+        raise RunnerManagerApplicationInstallError(_INSTALL_ERROR_MESSAGE) from err
+
+
+def _enable_service() -> None:
+    """Enable the github runner manager service.
+
+    Raises:
+        RunnerManagerApplicationStartupError: Unable to startup the service.
+    """
+    try:
+        systemd.service_enable(GITHUB_RUNNER_MANAGER_SERVICE_NAME)
+        if not systemd.service_running(GITHUB_RUNNER_MANAGER_SERVICE_NAME):
+            systemd.service_start(GITHUB_RUNNER_MANAGER_SERVICE_NAME)
+    except SystemdError as err:
+        raise RunnerManagerApplicationStartupError(_SERVICE_SETUP_ERROR_MESSAGE) from err
+
+
+def _setup_config_file(state: CharmState, app_name: str, unit_name: str) -> Path:
+    """Write the configuration to file.
+
+    Args:
+        state: The charm state.
+        app_name: The Juju application name.
+        unit_name: The Juju unit.
+    """
+    config = create_application_configuration(state, app_name, unit_name)
```

Comment:
Instead of passing the state from the charm to this module to the factory module, I'd rather generate the configuration in the charm and pass the configuration to this module